### ISSUE TITLE
Class and Instance dict should be an actual Dict

### ISSUE THREAD

**Issue by nhynes** [RustPython/RustPython#544]
Making Class and Instance `dict` be [a plain `HashMap`](https://github.com/RustPython/RustPython/blob/master/vm/src/pyobject.rs#L79) instead of an actual Dict [prevents adding keys to the class/instance `__dict__`](https://github.com/RustPython/RustPython/blob/master/vm/src/obj/objobject.rs#L179). The assumption of string keys is also incorrect:
```python
class A:
    pass

a = A()
a.__dict__[42] = 'val'
print(a.__dict__)  # cpython: {42: 'val'}
```

**Comment by windelbouwman** [comment]
Duplicates #46 . We have a history about this :). This appears very tricky. The good news is that an implementation is waiting to be connected to the dict object: https://github.com/RustPython/RustPython/blob/master/vm/src/dictdatatype.rs

**Comment by nhynes** [comment]
@windelbouwman very cool, thanks for the heads up! It looks like `PyAttributes` still needs to be addressed, but it shouldn't be too hard once `__hash__` is in place. I'll rename this issue.

**Comment by windelbouwman** [comment]
@nhynes I believe this issue can be closed now? Could you check on it?

**Comment by mtx2d** [comment]
Hi @windelbouwman , I am interested in the solution that addressed this issue and #46.

What is the PR that introduced proper dictionary implementation? Curious because from the [Video](https://www.youtube.com/watch?v=nJDY9ASuiLc) it says some help is needed for Python dict but I could not find the PR that helped with this feature.

**Comment by coolreader18** [comment]
@mtx2d Yep, and the implementation is still in `vm/src/dictdatatype.rs`. It was updated as well in #1875 and #2234, making it thread-safe and faster respectively.
